activerecord/lib/active_record/persistence.rb

```diff
@@ -1051,8 +1052,12 @@ def apply_scoping?(options)
         (self.class.default_scopes?(all_queries: true) || self.class.global_current_scope)
     end
 
-    def _primary_key_constraints_hash
-      { @primary_key => id_in_database }
+    def _query_constraints_hash
+      return { @primary_key => id_in_database } unless self.class.query_constraints_list
```

Comment:
This line is to keep backwards compatibility, eventually we expect it to go away once Rails starts deriving `query_constraints_list` from the database schema which means all existing models should get `["id"]` in the list "by convention"